File: objectModel/TypeScript/Cdm/CdmArgumentCollection.ts
import { isArray, isString } from 'util';
import {
    CdmArgumentDefinition,
    CdmCollection,
    CdmCorpusContext,
    cdmObjectType,
    CdmTraitReference
} from '../internal';

// tslint:disable: no-any
// tslint:disable: no-unsafe-any
export class CdmArgumentCollection extends CdmCollection<CdmArgumentDefinition> {
    protected get owner(): CdmTraitReference {
        return super.owner as CdmTraitReference;
    }

    protected set owner(value: CdmTraitReference) {
        super.owner = value;
    }

    constructor(ctx: CdmCorpusContext, owner: CdmTraitReference) {
        super(ctx, owner, cdmObjectType.argumentDef);
    }

    public push(arg: string | CdmArgumentDefinition | undefined, value?: any)
        : CdmArgumentDefinition {
        this.owner.resolvedArguments = false;
        const argument: CdmArgumentDefinition = super.push(arg);
        if (value !== undefined) {
            argument.value = value;
        }

        return argument;
    }

    public insert (index: number, arg: CdmArgumentDefinition): void {
        this.owner.resolvedArguments = false;
        super.insert(index, arg);
    }

    public fetchValue(name: string): any {
        for (const argument of this.allItems) {
            if (argument.name === name) {
                return argument.value;
            }
        }

        // special case with only one argument and no name give, make a big assumption that this is the one they want
        // right way is to look up parameter def and check name, but this public interface is for working on an unresolved def
        if (this.allItems.length === 1 && this.allItems[0].name === undefined) {
            return this.allItems[0].value;
        }

        return undefined;
    }

    public updateArgument(name: string, value: any): void {
        this.makeDocumentDirty();
        for (const argument of this.allItems) {
            if (argument.name === name) {
                argument.value = value;

                return;
            }
        }
        this.push(name, value);
    }
}

File: objectModel/Python/cdm/objectmodel/cdm_argument_collection.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Any, Union, TYPE_CHECKING

from cdm.enums import CdmObjectType

from .cdm_collection import CdmCollection

if TYPE_CHECKING:
    from .cdm_argument_def import CdmArgumentDefinition


class CdmArgumentCollection(CdmCollection):
    def __init__(self, ctx: 'CdmCorpusContext', owner: 'CdmTraitReference'):
        super().__init__(ctx, owner, CdmObjectType.ARGUMENT_DEF)

    def append(self, argument: Union[str, 'CdmArgumentDefinition'], value: Any = None) -> 'CdmArgumentDefinition':
        """Creates an argument based on the name and value provided and adds it to the collection."""

        from .cdm_argument_def import CdmArgumentDefinition

        argument = CdmArgumentDefinition(self.ctx, argument) if isinstance(argument, (str, type(None))) else argument
        if value is not None:
            argument.value = value

        self.owner._resolved_arguments = False
        return super().append(argument)

    def fetch_value(self, name: str):
        """
        Retrieves the value of the argument with the provided name.
        If no argument with provided name is found and there is only one argument with null name, retrieves the value of this argument.
        """

        for argument in self:
            if argument.name == name:
                return argument.value

        # special case with only one argument and no name give, make a big assumption that this is the one they want
        # right way is to look up parameter def and check name, but this public interface is for working on an unresolved def
        if len(self) == 1 and self[0].name is None:
            return self[0].value

        return None

    def insert(self, index: int, obj: 'CdmArgumentDefinition') -> None:
        self.owner._resolved_arguments = False
        return super().insert(index, obj)

    def update_argument(self, name: str, value: Any):
        """Updates the value of an existing argument, or, in case no argument is found with the provided name, an argument is created and added."""
        self._make_document_dirty()

        for argument in self:
            if argument.name == name:
                argument.value = value
                return
        self.append(name, value)
